webauthn-authenticator-rs/src/win10/gui.rs

```diff
@@ -0,0 +1,271 @@
+                // to MoveWindow later.
+                CreateWindowExW(
+                    WS_EX_TOPMOST | ex_style,
+                    WINDOW_CLASS,
+                    WINDOW_CLASS,
+                    style | WS_VISIBLE,
+                    CW_USEDEFAULT,
+                    CW_USEDEFAULT,
+                    1,
+                    1,
+                    parent,
+                    None,
+                    hinstance,
+                    None,
+                )
+            };
+            // trace!(?hwnd);
+
+            if hwnd == HWND(0) {
+                let e = unsafe { GetLastError() };
+                error!("window not created, {:?}", e);
+                sender.send(hwnd).ok();
+                return;
+            }
+
+            // Focus, foreground and reposition our window (if needed).
+            unsafe {
+                if !SetForegroundWindow(hwnd).as_bool() {
+                    trace!("Tried to set the foreground window, but the request was denied.");
+                }
+
+                if parent == HWND(0) {
+                    // When we have an un-parented window, make it invisible
+                    // and put it in the centre of the primary screen.
+                    SetLayeredWindowAttributes(hwnd, None, 0, LWA_ALPHA);
+                    Some((
+                        GetSystemMetrics(SM_CXSCREEN) / 2,
+                        GetSystemMetrics(SM_CYSCREEN) / 2,
+                    ))
+                } else {
+                    // When we have a parent window, MoveWindow is relative to the position
+                    // of the parent.
+                    get_window_rect(parent).map(half_size)
+                }
+                .map(|(x, y)| MoveWindow(hwnd, x, y, 1, 1, true));
+            }
+
+            // Now we can tell the main thread that the window is ready.
+            if sender.send(hwnd).is_err() {
+                return;
+            }
+
+            // Windows event loop
+            let mut msg: MSG = Default::default();
+            loop {
+                let res: bool = unsafe { GetMessageW(&mut msg, None, 0, 0) }.as_bool();
+                if !res {
+                    break;
+                }
+
+                if msg.message == WM_QUIT {
+                    unsafe {
+                        PostQuitMessage(msg.wParam.0 as i32);
+                    }
+                }
+                // trace!(?msg);
+                unsafe {
+                    TranslateMessage(&msg);
+                    DispatchMessageW(&msg);
+                }
+            }
+            // trace!("background stopped");
+        });
+
+        let hwnd = receiver.recv();
+        match hwnd {
+            Ok(HWND(0)) | Err(_) => Err(WebauthnCError::Internal),
+            Ok(hwnd) => Ok(Self { hwnd }),
+        }
+    }
+}
+
+impl Drop for Window {
+    fn drop(&mut self) {
+        // trace!("dropping window");
```

Comment:
Not that you need to change it, but a trickk in tracing is that you can "compile out" trace level calls so you don't need to comment them out in all cases :)

Reply:
ACK, I ended up commenting it out because it was noisy, but it was useful when writing it to make sure everything lived long enough, and the UI event loop was still running.